channeldb/payment_status.go

```diff
@@ -124,3 +124,44 @@ func (ps PaymentStatus) updatable() error {
 		return fmt.Errorf("%w: %v", ErrUnknownPaymentStatus, ps)
 	}
 }
+
+// Registrable returns an error to specify whether adding more HTLCs to the
+// payment with its current status is allowed. A payment can accept new HTLC
+// registration when it's newly created, or none of its HTLCs is in a terminal
+// state.
+func (ps PaymentStatus) Registrable(settled, paymentFailed bool) error {
+	switch ps {
+	// The payment has been created so we allow adding new HTLCs.
+	case StatusInitiated:
+		return nil
+
+	// There are still inflight HTLCs and we need to check whether there
+	// are settled HTLCs or the payment is failed.
+	case StatusInFlight:
+		// If we already have settled HTLCs, we won't allow adding more
+		// HTLCs.
+		if settled {
+			return ErrPaymentAlreadySucceeded
```

Comment:
Nice, so instead of returning `ErrPaymentTerminal` for both cases, we now have a more specific reason.